### PR TITLE
eBPF observability

### PR DESCRIPTION
**Pull request by nlgwcy**
**What type of PR is this?**

/kind feature

**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:
https://github.com/kmesh-net/kmesh/issues/320

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note

```

### CODE REVIEW

**Comment by hzxuzhonghu** [comment]
cc @LiZhenCheng9527

File: docs/proposal/ebpf-observability.md [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+creation-date: 2024-06-28
+
+
+---
+
+## eBPF Observability
+
+### Summary
+
+This proposal describes how to use the eBPF to implement observable information of Kmesh, and focuses on the design of logs and metrics.
```
Comment (LiZhenCheng9527) [inline docs/proposal/ebpf-observability.md:21]:
```suggestion
This proposal describes how to use the eBPF to implement observable information of Kmesh, and focuses on the design of access logs and metrics.

```

File: docs/proposal/ebpf-observability.md [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+
+### Motivation
+
+Like other service grid data planes, observability is an important function of the grid data plane. Kmesh needs to provide observation methods to help O&M personnel better understand the current network status.
```
Comment (LiZhenCheng9527) [inline docs/proposal/ebpf-observability.md:25]:
```suggestion
Like other service mesh data planes, observability is an important ability of the mesh data plane. Kmesh needs to provide observation methods to help O&M personnel better understand the current network status.

```

File: docs/proposal/ebpf-observability.md [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+
+#### Non-Goals
+
+NA
+
+### Proposal
+
+As designed in [observability](https://github.com/kmesh-net/kmesh/blob/main/docs/proposal/observability.md), the following information needs to be reported in eBPF:
+
+logs:
```
Comment (LiZhenCheng9527) [inline docs/proposal/ebpf-observability.md:35]:
```suggestion
**access logs:**
```

File: docs/proposal/ebpf-observability.md [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+
+When a connection is closed, the following informations about the connection need reported: duration, sent_bytes, received bytes, whether the connection is set up successfully, and the close time.
+
+metrics:
```
Comment (LiZhenCheng9527) [inline docs/proposal/ebpf-observability.md:39]:
```suggestion
**metrics:**

```

File: docs/proposal/ebpf-observability.md [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+
+Counts the number of connections established between a pair of IP addresses, number of connections closed, sent_bytes, received bytes, and number of connection establishment failures.
```
Comment (LiZhenCheng9527) [inline docs/proposal/ebpf-observability.md:41]:
```suggestion
Counts the number of connections established between a pair of IP, number of connections closed, sent_bytes, received bytes, and number of connection establishment failures.

```

File: bpf/kmesh/probes/probe.h [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+    // INGRESS senario
+    if (direction == INGRESS)
+        storage->connect_ns = bpf_ktime_get_ns();
+    storage->direction = direction;
+    storage->connect_success = true;
+
+    metric_on_connect(sk, tcp_sock, storage);
+}
+
+static inline void observe_on_close(struct bpf_sock *sk)
```
Comment (LiZhenCheng9527) [inline bpf/kmesh/probes/probe.h:50]:
```suggestion
static inline void observe_on_connect_close(struct bpf_sock *sk)
```

File: bpf/kmesh/probes/probe.h [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+    if (!storage) {
+        BPF_LOG(ERR, PROBE, "bpf_sk_storage_get failed\n");
+        return;
+    }
+
+    storage->connect_ns = bpf_ktime_get_ns();
+    return;
+}
+
+static inline void observe_on_connect(struct bpf_sock *sk, __u8 direction)
```
Comment (LiZhenCheng9527) [inline bpf/kmesh/probes/probe.h:23]:
```suggestion
static inline void observe_on_connect_open(struct bpf_sock *sk, __u8 direction)
```

File: bpf/kmesh/probes/metrics.h [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+{
+    struct metric_key key = {0};
+    struct metric_data data = {0};
+    struct metric_data *metric = NULL;
+
+    construct_metric_key(sk, &key);
+    metric = (struct metric_data *)bpf_map_lookup_elem(&map_of_metrics, &key);
+    if (!metric) {
+        // connect failed
+        data.conn_open++;
```
Comment (LiZhenCheng9527) [inline bpf/kmesh/probes/metrics.h:75]: Does conn.open need to be plus even if the link establishment fails?

File: bpf/kmesh/probes/metrics.h [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
+        data.direction = storage->direction;
+        data.conn_failed++;
+        int err = bpf_map_update_elem(&map_of_metrics, &key, &data, BPF_NOEXIST);
+        if (err)
+            BPF_LOG(ERR, PROBE, "metric_on_connect failed, err is %d\n", err);
```
Comment (LiZhenCheng9527) [inline bpf/kmesh/probes/metrics.h:80]:
```suggestion
            BPF_LOG(ERR, PROBE, "metric_on_connect update failed, err is %d\n", err);
```

File: bpf/kmesh/probes/metrics.h [hunk 10]
```diff
@@ -0,0 +1,91 @@
...
+    struct metric_data *metric = NULL;
+
+    construct_metric_key(sk, &key);
+    metric = (struct metric_data *)bpf_map_lookup_elem(&map_of_metrics, &key);
+    if (!metric) {
+        data.conn_open++;
+        data.direction = storage->direction;
+        int err = bpf_map_update_elem(&map_of_metrics, &key, &data, BPF_NOEXIST);
+        if (err)
+            BPF_LOG(ERR, PROBE, "metric_on_connect failed, err is %d\n", err);
```
Comment (LiZhenCheng9527) [inline bpf/kmesh/probes/metrics.h:55]:
```suggestion
            BPF_LOG(ERR, PROBE, "metric_on_connect update failed, err is %d\n", err);
```

File: bpf/kmesh/probes/metrics.h [hunk 11]
```diff
@@ -0,0 +1,91 @@
...
+static inline void
+metric_on_close(struct bpf_sock *sk, struct bpf_tcp_sock *tcp_sock, struct sock_storage_data *storage)
+{
+    struct metric_key key = {0};
+    struct metric_data data = {0};
+    struct metric_data *metric = NULL;
+
+    construct_metric_key(sk, &key);
+    metric = (struct metric_data *)bpf_map_lookup_elem(&map_of_metrics, &key);
+    if (!metric) {
```
Comment (LiZhenCheng9527) [inline bpf/kmesh/probes/metrics.h:73]: There must have been a map generated when the link was created, then not finding it when the link was closed. I think there's been an error.

File: bpf/kmesh/probes/metrics.h (hunk 8 above)
Comment (nlgwcy) [reply bpf/kmesh/probes/metrics.h:75]: ok, I'll fix it.

File: bpf/kmesh/probes/metrics.h (hunk 11 above)
Comment (nlgwcy) [reply bpf/kmesh/probes/metrics.h:73]: The metric record is created only when the connection is successfully established. If the connection fails to be established, the connection failure record needs to be recorded when the connection is closed.

File: bpf/kmesh/probes/probe.h (hunk 6 above)
Comment (nlgwcy) [reply bpf/kmesh/probes/probe.h:50]: just like `tcp_close` in kernel code, I think it no needs to modify.

File: bpf/kmesh/probes/probe.h (hunk 7 above)
Comment (nlgwcy) [reply bpf/kmesh/probes/probe.h:23]: `observe_on_connect_established` may be more suitable.

File: bpf/kmesh/probes/metrics.h [hunk 12]
```diff
@@ -0,0 +1,91 @@
...
+#define __KMESH_BPF_METRICS_H__
+#include "bpf_common.h"
+
+// metrics
+struct metric_key {
+    struct ip_addr src_ip;
+    struct ip_addr dst_ip;
+};
+
+struct metric_data {
```
[the remark this answers is not in the document]
Comment (nlgwcy) [reply bpf/kmesh/probes/metrics.h:11]: yes. This feature will be supported later.

File: bpf/kmesh/probes/access_log.h [hunk 13]
```diff
@@ -0,0 +1,69 @@
+#ifndef __KMESH_BPF_ACCESS_LOG_H__
+#define __KMESH_BPF_ACCESS_LOG_H__
+
+#include "bpf_common.h"
+
+// access log
+enum {
+    INVALID_DIRECTION = 0,
```
Comment (YaoZengzeng) [inline bpf/kmesh/probes/access_log.h:8]: This doesn't seem to be used, is it necessary?

File: bpf/kmesh/probes/access_log.h [hunk 14]
```diff
@@ -0,0 +1,69 @@
...
+    }
+    return;
+}
+
+static inline void
+report_access_log(struct bpf_sock *sk, struct bpf_tcp_sock *tcp_sock, struct sock_storage_data *storage)
+{
+    struct access_log *log = NULL;
+
+    // strore tuple
```
Comment (YaoZengzeng) [inline bpf/kmesh/probes/access_log.h:51]:
```suggestion
    // store tuple
```

typo?

File: docs/proposal/ebpf-observability.md [hunk 15]
```diff
@@ -0,0 +1,108 @@
...
+
+As designed in [observability](https://github.com/kmesh-net/kmesh/blob/main/docs/proposal/observability.md), the following information needs to be reported in eBPF:
+
+**access logs:**
+
+When a connection is closed, the following informations about the connection need reported: duration, sent_bytes, received bytes, whether the connection is set up successfully, and the close time.
+
+**metrics:**
+
+Counts the number of connections established between a pair of IP, number of connections closed, sent_bytes, received bytes, and number of connection establishment failures.
```
Comment (YaoZengzeng) [inline docs/proposal/ebpf-observability.md:41]:
```suggestion
Counts the number of connections established between two IPs, number of connections closed, sent_bytes, received bytes, and number of connection establishment failures.

```

File: bpf/kmesh/probes/probe.h [hunk 16]
```diff
@@ -0,0 +1,69 @@
...
+
+    storage->connect_ns = bpf_ktime_get_ns();
+    return;
+}
+
+static inline void observe_on_connect_established(struct bpf_sock *sk, __u8 direction)
+{
+    struct bpf_tcp_sock *tcp_sock = NULL;
+    struct sock_storage_data *storage = NULL;
+    __u64 flags = (direction == EGRESS) ? 0 : BPF_LOCAL_STORAGE_GET_F_CREATE;
```
Comment (hzxuzhonghu) [inline bpf/kmesh/probes/probe.h:27]: what is it for?

File: bpf/kmesh/probes/access_log.h [hunk 17]
```diff
@@ -0,0 +1,69 @@
+#ifndef __KMESH_BPF_ACCESS_LOG_H__
+#define __KMESH_BPF_ACCESS_LOG_H__
+
+#include "bpf_common.h"
+
+// access log
+enum {
+    INVALID_DIRECTION = 0,
+    INGRESS = 1,
+    EGRESS = 2,
```
Comment (hzxuzhonghu) [inline bpf/kmesh/probes/access_log.h:10]: INBOUND
OUTBOUND

File: bpf/kmesh/probes/probe.h (hunk 16 above)
Comment (hzxuzhonghu) [reply bpf/kmesh/probes/probe.h:27]: Cannot we always use `BPF_LOCAL_STORAGE_GET_F_CREATE `

Comment (nlgwcy) [reply bpf/kmesh/probes/probe.h:27]: This is a user-defined data area associated with sock. You need to set BPF_LOCAL_STORAGE_GET_F_CREATE flag when creating the data area for the first time.

File: bpf/kmesh/probes/metrics.h [hunk 18]
```diff
@@ -0,0 +1,116 @@
...
+        key->src_ip.ip4 = sk->src_ip4;
+        key->dst_ip.ip4 = sk->dst_ip4;
+    } else {
+        IP6_COPY(key->src_ip.ip6, sk->src_ip6);
+        IP6_COPY(key->dst_ip.ip6, sk->dst_ip6);
+    }
+    return;
+}
+
+static inline void metric_notify(struct bpf_sock *sk)
```
[the remark this answers is not in the document]
Comment (nlgwcy) [reply bpf/kmesh/probes/metrics.h]: fixed

File: bpf/kmesh/probes/access_log.h (hunk 17 above)
Comment (nlgwcy) [reply bpf/kmesh/probes/access_log.h]: fixed

File: bpf/kmesh/probes/probe.h (hunk 16 above)
Comment (hzxuzhonghu) [reply bpf/kmesh/probes/probe.h]: I think `BPF_LOCAL_STORAGE_GET_F_CREATE ` can be always set even for reading the data

File: bpf/kmesh/workload/cgroup_sock.c [hunk 19]
```diff
@@ -81,6 +82,8 @@ int cgroup_connect4_prog(struct bpf_sock_addr *ctx)
         // if tail call failed will run this code
         BPF_LOG(ERR, KMESH, "workload tail call failed, err is %d\n", ret);
     }
+
+    observe_on_pre_connect(ctx->sk);
```
[the remark this answers is not in the document]
Comment (hzxuzhonghu) [reply bpf/kmesh/workload/cgroup_sock.c:86]: how do you see that

File: bpf/include/bpf_common.h [hunk 20]
```diff
@@ -49,6 +49,19 @@ struct {
     __uint(map_flags, BPF_F_NO_PREALLOC);
 } map_of_manager SEC(".maps");
 
+struct sock_storage_data {
+    __u64 connect_ns;
```
Comment (hzxuzhonghu) [inline bpf/include/bpf_common.h:53]: connect_at or connect_time

File: bpf/include/bpf_common.h [hunk 21]
```diff
@@ -49,6 +49,19 @@ struct {
...
+    __u8 direction;
+    __u8 connect_success;
+};
+
+struct {
+    __uint(type, BPF_MAP_TYPE_SK_STORAGE);
+    __uint(map_flags, BPF_F_NO_PREALLOC);
+    __type(key, int);
+    __type(value, struct sock_storage_data);
+} map_of_sock_storage SEC(".maps");
```
Comment (hzxuzhonghu) [inline bpf/include/bpf_common.h:63]: map_of_sock_info

File: bpf/include/bpf_common.h [hunk 22]
```diff
@@ -49,6 +49,19 @@ struct {
     __uint(map_flags, BPF_F_NO_PREALLOC);
 } map_of_manager SEC(".maps");
 
+struct sock_storage_data {
```
Comment (hzxuzhonghu) [inline bpf/include/bpf_common.h:52]: sock_info

File: bpf/include/bpf_log.h [hunk 23]
```diff
@@ -16,6 +16,7 @@
 #define BPF_LOGTYPE_SOCKOPS BPF_DEBUG_OFF
 #define BPF_LOGTYPE_XDP     BPF_DEBUG_OFF
 #define BPF_LOGTYPE_SENDMSG BPF_DEBUG_OFF
+#define BPF_LOGTYPE_PROBE   BPF_DEBUG_ON
```
Comment (hzxuzhonghu) [inline bpf/include/bpf_log.h:19]: Suppose you will change to debug before merge

File: bpf/kmesh/probes/access_log.h [hunk 24]
```diff
@@ -0,0 +1,69 @@
...
+enum {
+    INVALID_DIRECTION = 0,
+    INBOUND = 1,
+    OUTBOUND = 2,
+};
+
+struct access_log {
+    struct bpf_sock_tuple tuple;
+    __u64 duration; // ns
+    __u64 close_ns;
```
Comment (hzxuzhonghu) [inline bpf/kmesh/probes/access_log.h:16]: @LiZhenCheng9527 I am not sure do we need close time

File: bpf/kmesh/probes/metrics.h [hunk 25]
```diff
@@ -0,0 +1,116 @@
...
+};
+
+#define MAP_SIZE_OF_METRICS 100000
+struct {
+    __uint(type, BPF_MAP_TYPE_HASH);
+    __type(key, struct metric_key);
+    __type(value, struct metric_data);
+    __uint(max_entries, MAP_SIZE_OF_METRICS);
+    __uint(map_flags, BPF_F_NO_PREALLOC);
+} map_of_metrics SEC(".maps");
```
Comment (hzxuzhonghu) [inline bpf/kmesh/probes/metrics.h:27]: who is responsible for the delete of this map?

File: bpf/kmesh/workload/cgroup_sock.c (hunk 19 above)
[the remark this answers is not in the document]
Comment (hzxuzhonghu) [reply bpf/kmesh/workload/cgroup_sock.c:86]: I see what you mean, it doesnot need to record connect time that in tail call.

**Pull request merged**